stdlib/dynarray.mli

```diff
@@ -0,0 +1,190 @@
+(*                                                                        *)
+(**************************************************************************)
+
+(** Growable, mutable array.
+
+
+
+    @since 5.1
+*)
+
+type 'a t
+(** A dynamic array containing values of type ['a].
+
+    This contains an underlying {!array} along with a size.
+    Operations such as {!push}, {!append}, and {!append_seq}, extend the
+    size (and might reallocate the underlying array).
+
+    Operations such as {!pop}, and {!truncate}, reduce the size. *)
+
+val create : unit -> 'a t
+(** Create a new, empty array. *)
+
+val make : int -> 'a -> 'a t
+(** [make n x] makes a array of size [n], filled with [x]. *)
+
+val init : int -> (int -> 'a) -> 'a t
+(** Init the array with the given function and size. *)
+
+val clear : 'a t -> unit
+(** Clear the content of the array.
+    This ensures that [length v = 0] but the underlying array is kept,
+    and possibly references to former elements, which are therefore
+    not garbage collectible. *)
+
+val ensure_capacity_with : 'a t -> filler:'a -> int -> unit
+(** Make sure that the array has at least the given capacity (underlying size).
+
+    This is a more advanced operation that is only useful for performance
+    purposes.
+
+    @param filler an element used if the underlying array is empty,
+      to initialize it. It will be retained until the array is totally
+      empty or until it is garbage collected.
+    @raise Invalid_arg if the size is not suitable (negative, or too big for
+    OCaml arrays)
+*)
+
+val ensure_capacity_nonempty : 'a t -> int -> unit
+(** Make sure that the array has at least the given capacity (underlying size),
+    assuming it is non-empty. The first element is used as the filler.
+
+    This is a more advanced operation that is only useful for performance
+    purposes.
+
+    @raise Invalid_arg if the array is empty or
+      if the size is not suitable (negative, or too big for OCaml arrays)
+*)
+
+val is_empty : 'a t -> bool
+(** Is the array empty? This is synonymous to [length a = 0]. *)
+
+val push : 'a t -> 'a -> unit
+(** Add an element at the end of the array. This might extend the underlying
+    array if it is full.
+
+    Calling [push] [n] times is amortized O(n) complexity,
+    and O(ln(n)) reallocations of the underlying array. *)
+
+val unsafe_push : 'a t -> 'a -> unit
+(** Push an element, assuming there is capacity for it
+    (e.g. using {!ensure_capacity}).
+
+    It is unspecified what happens if the capacity is not enough.
+    This is for advanced used only. *)
+
+val append : 'a t -> 'a t -> unit
+(** [append a b] adds all elements of [b] to [a]. [b] is not modified. *)
+
+val append_array : 'a t -> 'a array -> unit
+(** Like {!append}, with an array. *)
+
+val append_seq : 'a t -> 'a Seq.t -> unit
+(** Like {!append} but with an iterator. *)
+
+val append_list : 'a t -> 'a list -> unit
+(** Like {!append} but with a list. *)
```

Comment:
I don't think all these specialized `append_{array,seq,list}` scale well. I may want to append a bigarray, a set, my own iterable datastructure etc. I would suggest replacing them with:

```ocaml
val append_iter : 'a t -> (('a -> unit) -> 'b -> unit) -> 'b -> unit
```

I could see the point of making an exception for `append_array` for bliting reasons, but other than that writing: 

```ocaml
Dynarray.append_iter List.iter l
Dynarray.append_iter Seq.iter s
```
rather than
```ocaml
Dynarray.append_list l
Dynarray.append_seq s
```
Doesn't feel egregious.

Reply:
... then people maybe would give us a break about the performance of `Seq` ;-)

Reply:
Well that type wouldn't work with any of the stdlib's iter functions. So I'm not sure what it would bring here.

> ... then people maybe would give us a break about the performance of Seq ;-)

If that was the only problem. The gripe is rather hiding effects and holding ressources behind lazy computations…

Reply:
I added `append_iter`.